### PR TITLE
Docker for mac support

### PR DESCRIPTION
**Pull request by lionelvillard**
This change is a continuation of PR #2440 and PR #2452

### CODE REVIEW

**Comment by rabbah** [comment]
worked for me!

@csantanapr @style95 FYI.
@ddragosd your wish is closer to becoming true.

**Comment by rabbah** [comment]
Without the nuke consul PR I failed at the consul deployment phase which seems to require sudo - not going to bother debugging this and instead will wait for that PR to be merged.

**Comment by rabbah** [comment]
@lionelvillard do we need to add this to the docs https://github.com/apache/incubator-openwhisk/pull/2440#issuecomment-312435272?

**Comment by lionelvillard** [comment]
@rabbah I added this to the doc: https://github.com/apache/incubator-openwhisk/pull/2524/commits/0ff4400512cda733b0397e5f6f5771cb177ce999#diff-5a5d069dcad99a9f8109ac77d1b9860cR258

I think that's the proper way when using `brew`

File: ansible/README.md [hunk 1]
```diff
@@ -27,19 +27,6 @@ You need to depends on the workarounds until Docker provides official methods.
...
-Currently however, Docker for Mac does not support such a feature.
-
-There are many workarounds for this.
-One way is to use `socat` command to setup proxy for the UNIX socket.
-
-```
-docker run -d -v /var/run/docker.sock:/var/run/docker.sock -p 4243:2375 bobrik/socat TCP4-LISTEN:2375,fork,reuseaddr UNIX-CONNECT:/var/run/docker.sock
-```
-
-If you want to deploy OpenWhisk in distributed environment, you are required to enable Docker remote API in all Mac hosts.
```
Comment (csantanapr) [inline ansible/README.md]: Also nuke the section **Setup proxy container to run unit tests (optional)**
No need it anymore

**Comment by csantanapr** [comment]
@rabbah I just tested this PR and it worked 🎉 🎉 🎉 🎉 🎉

File: ansible/README.md (hunk 1 above)
Comment (style95) [inline ansible/README.md]: With these changes, could we communicate with action container using container IP such as `172.17.0.10` ?

**Comment by rabbah** [comment]
@csantanapr there are still references to port 4243 after this pr; is the goal to remove all those?

**Comment by rabbah** [comment]
i cannot not run the container tests yes - no proxy deployed @csantanapr @style95

**Comment by rabbah** [comment]
having a couple of sudo hiccups that need investigating:

```
TASK [prereq : install requests] ******************************************************************************************************************************************************************************************************************************************************************
Thursday 27 July 2017  12:25:01 -0400 (0:00:00.022)       0:00:01.242 ********* 
fatal: [172.17.0.1]: FAILED! => {"changed": false, "failed": true, "module_stderr": "sudo: a password is required\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```

```
TASK [controller : remove controller log directory] ***********************************************************************************************************************************************************************************************************************************************
Thursday 27 July 2017  12:39:10 -0400 (0:00:00.254)       0:00:02.178 ********* 
fatal: [172.17.0.1]: FAILED! => {"changed": false, "failed": true, "module_stderr": "sudo: a password is required\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```

**Comment by rabbah** [comment]
@jeremiaswerner @markusthoemmes why do we have ansible privilege escalation? https://github.com/apache/incubator-openwhisk/search?utf8=%E2%9C%93&q=become+&type=

i'd like to remove it all (if it's truly needed for some deployments, then i can guard it so it's not enable for local development)

**Comment by rabbah** [comment]
git blames for the becomes point to many fingers - i'm wondering if we really need these.
perhaps in some environments it might be required, in which case i'd exempt all local deployments if possible.

File: ansible/README.md (hunk 1 above)
Comment (rabbah) [inline ansible/README.md]: @style95 is correct, the proxy (or some other solution) is still required.

**Comment by csantanapr** [comment]
@rabbah yeah if possible I would eliminate the need to have docker engine expose via tcp on port 4243, this is not the default way docker is install and causing problems with people running locally.

port 4243 would be required for docker-machine only.

For the tests using 4243 we could look at them to see if they can a property from whisk.properties or an environment variable to indicate to use the flag to use port 4243, if not then assume docker-cli uses the unix-socket device.

File: ansible/README.md (hunk 1 above)
Comment (csantanapr) [inline ansible/README.md]: sorry, about that leave the proxy info as required for testing for now :-)
We should look into our tests to see how it can be optional, on docker for mac the docker host is not need it and we should adjust things like this:
https://github.com/apache/incubator-openwhisk/blob/master/tests/src/test/scala/actionContainers/ActionContainer.scala#L89
```
    private lazy val dockerCmd: String = {
        val hostStr = if (WhiskProperties.onMacOSX()) {
            s" --host tcp://${WhiskProperties.getMainDockerEndpoint()} "
        } else {
            " "
        }
        s"$dockerBin $hostStr"
    }
```

Comment (rabbah) [inline ansible/README.md]: ideally this goes away but need either the proxy to tap... no free lunch yet.

Comment (csantanapr) [inline ansible/README.md]: Also like this ping test, I think this tests are now move to ansible, this was required when we use this tests that system was deployed correctly, now ansible serves that role.
https://github.com/apache/incubator-openwhisk/blob/master/tests/src/test/scala/services/PingTests.scala#L95
```
    it should "check that the main docker endpoint is functioning" in {
        pingDocker("main.docker.endpoint", WhiskProperties.getMainDockerEndpoint)
    }
```

File: ansible/environments/mac/group_vars/all [hunk 2]
```diff
@@ -4,6 +4,8 @@ whisk_logs_dir: /Users/Shared/wsklogs
 docker_registry: ""
 docker_dns: ""
 
+ansible_python_interpreter: "/usr/local/bin/python"
```
Comment (rabbah) [inline ansible/environments/mac/group_vars/all]: this environment should be deleted then.

File: ansible/README.md (hunk 1 above)
Comment (rabbah) [inline ansible/README.md]: we should back out this change initially - the 4243 issue is a bit messy but in general this pr makes an improvement; we should try to get it in incrementally.

File: ansible/README.md [hunk 3]
```diff
@@ -72,8 +59,9 @@ systemProp.http.proxyPort=3128
 **Caveat:** All Ansible commands are meant to be executed from the `ansible` directory.
 This is important because that's where `ansible.cfg` is located which contains generic settings that are needed for the remaining steps.
 
-In all instructions, replace `<environment>` with your target environment. e.g. `mac`if you want to deploy using a local mac setup.
+In all instructions, replace `<environment>` with your target environment. e.g. `docker-machine`if you want to deploy using a local docker-machine setup.
 By default, if you omit the `-i` parameter, the `local` environment will be used.
+If using Docker for Mac use the `local` environment.
```
Comment (rabbah) [inline ansible/README.md]: there may be other places to clean up docs for mac - worth double checking.

File: ansible/roles/controller/tasks/deploy.yml [hunk 4]
```diff
@@ -14,7 +14,6 @@
     path: "{{ whisk_logs_dir }}/controller{{ groups['controllers'].index(inventory_hostname) }}"
     state: directory
     mode: 0777
-  become: true
```
Comment (rabbah) [inline ansible/roles/controller/tasks/deploy.yml]: in general there are many of these there are a problem - addressed in #2536 since it seems wrong to begin with.

File: ansible/roles/invoker/tasks/deploy.yml [hunk 5]
```diff
@@ -26,34 +26,71 @@
   retries: "{{ docker.pull.retries }}"
   delay: "{{ docker.pull.delay }}"
 
-- name: determine docker root dir
+- name: "determine docker root dir on docker-machine"
   uri:  url="http://{{ inventory_hostname }}:{{ docker.port }}/info" return_content=yes
-  register: dockerInfo
+  register: dockerInfo_output
+  when: whisk_version_name == "mac"
```
Comment (rabbah) [inline ansible/roles/invoker/tasks/deploy.yml]: are these all wrong now? should they be "local"?

File: tools/build/redo [hunk 6]
```diff
@@ -261,6 +261,10 @@ Components = [
                   'install catalog',
                   yaml = 'postdeploy.yml'),
 
+    makeComponent('apigateway',
```
Comment (rabbah) [inline tools/build/redo]: let's pull this out into a separate pr.

File: ansible/roles/controller/tasks/deploy.yml (hunk 4 above)
Comment (csantanapr) [inline ansible/roles/controller/tasks/deploy.yml]: this is clean now

File: ansible/roles/invoker/tasks/deploy.yml (hunk 5 above)
Comment (csantanapr) [inline ansible/roles/invoker/tasks/deploy.yml]: this is still required only for while we have docker-machine support

File: tools/build/redo (hunk 6 above)
Comment (csantanapr) [inline tools/build/redo]: will do

File: ansible/roles/invoker/tasks/deploy.yml (hunk 5 above)
Comment (rabbah) [inline ansible/roles/invoker/tasks/deploy.yml]: ok so this condition is wrong? instead of == "mac" it should be == "docker-machine" you're saying?

File: ansible/roles/invoker/tasks/deploy.yml [hunk 7]
```diff
@@ -26,9 +26,25 @@
...
+- name: "determine docker root dir"
+  shell: echo -e "GET http:/v1.24/info HTTP/1.0\r\n" | nc -U /var/run/docker.sock | grep "{"
+  args:
+    executable: /bin/bash
+  register: dockerInfo_output
+  when: whisk_version_name != "mac"
+
+- set_fact:
+    dockerInfo: "{{ dockerInfo_output.stdout|from_json }}"
+  when: whisk_version_name != "mac"
```
Comment (rabbah) [inline ansible/roles/invoker/tasks/deploy.yml]: what about this condition? "mac" is gone so is this always done?

File: ansible/roles/invoker/tasks/deploy.yml [hunk 8]
```diff
@@ -39,20 +55,42 @@
...
+    INFO=`echo -e "GET http:/v1.24/containers/json?filters={{ '{"name":[ "invoker" ]}' | urlencode }} HTTP/1.0\r\n" | nc -U /var/run/docker.sock | grep "{"`
+    if [ -z "$INFO" ]; then
+      echo []
+    else
+      echo $INFO
+    fi
+  args:
+    executable: /bin/bash
+  register: invokerInfo_output
+  when: whisk_version_name != "mac"
```
Comment (rabbah) [inline ansible/roles/invoker/tasks/deploy.yml]: and here.

File: ansible/roles/invoker/tasks/deploy.yml [hunk 9]
```diff
@@ -39,20 +55,42 @@
...
+
+- set_fact:
+    invokerInfo: "{{ invokerInfo_output.stdout|from_json }}"
+  when: whisk_version_name != "mac"
```
Comment (rabbah) [inline ansible/roles/invoker/tasks/deploy.yml]: and here.

File: ansible/roles/invoker/tasks/deploy.yml (hunk 5 above)
Comment (csantanapr) [inline ansible/roles/invoker/tasks/deploy.yml]: whisk_version_name is only set to `mac` in ansible/environment/mac/group_vars/all

File: ansible/README.md [hunk 10]
```diff
@@ -297,7 +297,7 @@ without spaces as there is no current fix available to this problem.
 
 #### Changing limits
 The system throttling limits can be changed by modifying the `group_vars` for your environment. For example,
-mac users will find the limits in this file [./environments/mac/group_vars/all](./environments/mac/group_vars/all):
+Docker for Mac or ubuntu users will find the limits in this file [./environments/local/group_vars/all](./environments/local/group_vars/all):
```
Comment (rabbah) [inline ansible/README.md]: Ubuntu (capital U).

File: ansible/README.md (hunk 3 above)
Comment (csantanapr) [inline ansible/README.md]: just found one additional place to update

File: ansible/logs.yml [hunk 11]
```diff
@@ -27,10 +27,10 @@
 - hosts: shared,core,invokers
   tasks:
   - name: get all docker containers
-    local_action: shell docker --host tcp://{{inventory_hostname}}:{{docker.port}} ps -a --format="{% raw %}{{.Names}}{% endraw %}"
```
Comment (rabbah) [inline ansible/logs.yml]: will this work for a distributed deployment?

File: ansible/README.md [hunk 12]
```diff
@@ -283,7 +297,7 @@ without spaces as there is no current fix available to this problem.
...
```
Comment (rabbah) [inline ansible/README.md]: Ubuntu (capital U).

File: ansible/logs.yml (hunk 11 above)
Comment (rabbah) [inline ansible/logs.yml]: i'm not sure we need this change.

**Comment by csantanapr** [comment]
continue work in #2686
